app/src/org/commcare/dalvik/activities/CommCareHomeActivity.java

```diff
@@ -210,7 +217,116 @@ private void processFromLoginLaunch() {
+                getIntent().getStringExtra(LoginActivity.LOGIN_MODE));
+
+        if (loginMode == LoginActivity.LoginMode.PRIMED) {
+            launchPinCreateScreen(loginMode);
+        } else if (loginMode == LoginActivity.LoginMode.PASSWORD) {
+            boolean pinCreationEnabledForApp = DeveloperPreferences.shouldOfferPinForLogin();
+            if (!pinCreationEnabledForApp) {
+                return;
+            }
+
+            boolean userManuallyEnteredPasswordMode = getIntent()
+                    .getBooleanExtra(LoginActivity.MANUAL_SWITCH_TO_PW_MODE, false);
+            boolean alreadyDismissedPinCreation =
+                    CommCareApplication._().getCurrentApp().getAppPreferences()
+                            .getBoolean(CommCarePreferences.HAS_DISMISSED_PIN_CREATION, false);
+            if (!alreadyDismissedPinCreation || userManuallyEnteredPasswordMode) {
+                showPinChoiceDialog(loginMode);
+            }
+        }
+    }
+
+    private void showPinChoiceDialog(final LoginActivity.LoginMode loginMode) {
+        String promptMessage;
+        if (CommCareApplication._().getRecordForCurrentUser().hasPinSet()) {
+            promptMessage = Localization.get("pin.dialog.prompt.reset");
+        } else {
+            promptMessage = Localization.get("pin.dialog.prompt.set");
         }
+
+        final PaneledChoiceDialog dialog = new PaneledChoiceDialog(this, promptMessage);
+
+        DialogChoiceItem createPinChoice = new DialogChoiceItem(
+                Localization.get("pin.dialog.yes"), -1, new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        dialog.dismiss();
+                        launchPinCreateScreen(loginMode);
+                    }
+                });
+
+        DialogChoiceItem nextTimeChoice = new DialogChoiceItem(
+                Localization.get("pin.dialog.not.now"), -1, new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        dialog.dismiss();
+                    }
+                });
+
+        DialogChoiceItem notAgainChoice = new DialogChoiceItem(
+                Localization.get("pin.dialog.never"), -1, new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        dialog.dismiss();
+                        CommCareApplication._().getCurrentApp().getAppPreferences()
+                                .edit()
+                                .putBoolean(CommCarePreferences.HAS_DISMISSED_PIN_CREATION, true)
+                                .commit();
+                        showPinFutureAccessDialog();
+                    }
+                });
+
+
+        dialog.setChoiceItems(new DialogChoiceItem[]{createPinChoice, nextTimeChoice, notAgainChoice});
+        dialog.addCollapsibleInfoPane(Localization.get("pin.dialog.extra.info"));
```

Comment:
User can click on a small icon button to see an explanation of what it means to set a PIN